Allow toggling of sim behaviour

ptsOSL opened the pull request:
Waiting for the diamond2-conversion merge into main to happen first.

ptsOSL committed:
Allow toggling of additional simulator behaviour

Allow selecting linopt function, disabling chromaticity aand
disabling radiation

ptsOSL committed:
Add tests for new sim params

Also fix some tests which were broken by reordering of
function arguments

ptsOSL committed:
Remove support for python 3.12 and 3.13 due to cothread issues

The CI is failing after python 3.11 due to a cothread seg
fault. These version of cothread are known to be unreliable so,
remove support for python 3.12 and 3.13 until after we have
switched from cothread to asyncio

ptsOSL commented:
I had to remove support for python 3.12 and python 3.13 as the tests were causing cothread to segfault. As we will shortly be switching to asyncio I think it is better to just stop supporting these python versions rather than figuring out what is causing the seg fault. After the switch we can support them again.

ptsOSL committed:
Refactor sim params into a dataclass

There will be a corresponding update to Virtac

ptsOSL committed:
Dont error on sphinx nitpicky warning

ptsOSL committed:
Adjust pyAT sim param setting

We now use enable_6d instead of the misleading radiation_on,
enable_6d is only called for linopt6. Also modify the validation
exceptions to be more accurate

MJGaughran commented on src/atip/simulator.py:
```diff
@@ -12,6 +13,34 @@
+    chromaticity: bool = True
+    radiation: bool = True
+
+    def __post_init__(self):
```
If you want to use __post_init__ for validation, you need to make sure the dataclass is immutable, i.e. frozen=True
```
@dataclass(frozen=True)
```

ptsOSL replied:
Added thanks

ptsOSL committed:
Make dataclass frozen

This will mean that post_init is called if anyone later modifies
the attributes of the dataclass. This ensure validation always

ptsOSL committed:
Move linopt validation check to dataclass

ptsOSL committed:
Split up disable_emittance flag tests

Now that disabling emittance on an existing ATSimulator isnt
supported, it makes more sense to have two seperate tests
one which tests a simulator with and one without emittance
disabled. Rather than testing both at once

MJGaughran reviewed (approved):
I don't think the tests are perfect, but after discussing this seems like a reasonable improvement over the original ones.

It would be good to distinguish between coverage and other failures in the CI, but I don't know how much of a pain that is with the copier template. I'll make sure to double check cause of failures in instances like this, in future.
